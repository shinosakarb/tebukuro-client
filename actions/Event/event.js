// @flow
import { createAction } from 'redux-actions'
import Router from 'next/router'
import type { Dispatch } from 'redux'

import ActionsType from '../../constants/Actions'
import type { EventProps } from '../../types/Event'
// TODO: Enable ESLint and flow check after API implemented.
/* eslint-disable */
// flow-disable-nextline
import EventAPI from '../../api/Event'

const create = createAction(ActionsType.Event.createEvent)

export const createEvent = (params: EventProps) => (dispatch: Dispatch) => (
  EventAPI.create(params)
    .then((json) => {
      Router.replace(`/event/${json.id}`)
      return dispatch(create(json))
    })
    .catch(error => dispatch(create(error)))
)
