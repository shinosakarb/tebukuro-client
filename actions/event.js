// @flow
import { createAction } from 'redux-actions'
import Router from 'next/router'
import type { Dispatch } from 'redux'
import ActionsType from '../constants/Actions'
// TODO: Enable ESLint and flow check after API implemented.
/* eslint-disable */
// flow-disable-nextline
import EventAPI from '../api/Event'
import type { EventProps } from '../types/Event'


const create = createAction(ActionsType.Event.createEvent, EventAPI.create)

// TODO: Return reject in thunk.
export const createEvent = (params: EventProps) => (dispatch: Dispatch, getState: Function) => {
    return dispatch(create(params)).then(res => {
       const id = getState().event.id
       !res.error && Router.replace(`/event/show?id=${id}`, `/event/${id}`)
    })
}

export const fetchEvent = createAction(ActionsType.Event.fetchEvent, EventAPI.find)
