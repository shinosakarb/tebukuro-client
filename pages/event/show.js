// @flow
import React from 'react'

export default () => (
  <div>
    <h3>This is the event page!</h3>
  </div>
)
