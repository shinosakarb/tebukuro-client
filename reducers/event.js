// @flow
import { handleActions } from 'redux-actions'
import Actions from '../constants/Actions'

export const eventInitialState = {}

const eventReducerMap = {
  [Actions.Event.createEvent]: {
    next: (state, action) => action.payload,
    throw: (state, action) => (
      { errors: action.payload.errors }
    ),
  },

  [Actions.Event.fetchEvent]: {
    next: (state, action) => action.payload,
    throw: (state, action) => (
      { errors: action.payload.errors }
    ),
  },
}

export default handleActions(eventReducerMap, eventInitialState)
