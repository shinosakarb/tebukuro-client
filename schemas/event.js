import { schema } from 'normalizr'
import ParticipantSchema from './participant'

const EventSchema = new schema.Entity('event', {
  participants: [ParticipantSchema],
})

export default EventSchema
