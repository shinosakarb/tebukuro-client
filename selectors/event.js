// @flow
import { createSelector } from 'reselect'

const getEventEntities = state => state.event.get('entities')
const getEventErrors = state => state.event.get('errors')

export const getEventId = (state: Object) => state.event.get('entityId')
export const getEventErrorsArray = createSelector(getEventErrors, errors => errors.toArray())
export const getCurrentEvent = createSelector(
  [getEventId, getEventEntities],
  (id, entities) => {
    const entitiesObject = entities.get(id.toString()).toObject()
    return {
      ...entitiesObject,
      participants: entitiesObject.participants.toArray(),
    }
  },
)
export const getHasWaitlist =
  createSelector(getCurrentEvent, event => event.participants.length >= event.quota)
export const getHasNotFoundError =
  createSelector(getEventErrors, errors => errors.indexOf('Not Found') !== -1)
