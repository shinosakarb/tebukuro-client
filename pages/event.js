// @flow
import React from 'react'
import Event from '../containers/Event'
import type { EventId } from '../types/Event'

export default (props: {url: {query: EventId}}) => (
  <div>
    <h3>This is the event page!</h3>
    <Event eventId={props.url.query.id} />
  </div>
)
