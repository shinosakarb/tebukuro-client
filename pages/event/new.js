// @flow
import React from 'react'
import withRedux from 'next-redux-wrapper'
import createStore from '../../store'
import { createEvent } from '../../actions/event'
import EventFormComponent from '../../components/EventForm'

type EventFormProps = {
  createEvent: Function
}

export const NewEvent = (props: EventFormProps) => (
  <div>
    <h3>This is the event form page!</h3>
    <EventFormComponent onSubmit={props.createEvent} />
  </div>
)

const mapDispatchToProps = { createEvent }

const connectProps = {
  createStore,
  mapDispatchToProps,
}

export default withRedux(connectProps)(NewEvent)
