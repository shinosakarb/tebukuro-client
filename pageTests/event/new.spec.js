import { shallow } from 'enzyme'
import shallowToJson from 'enzyme-to-json'
import { NewEvent } from '../../pages/event/new'

jest.mock('../../components/EventForm', () => 'EventFormComponent')
jest.mock('../../actions/event', () => ({
  createEvent: () => {},
}))

const testProps = {
  createEvent: jest.fn(),
}

describe('NewEvent', () => {
  it('renders the new event page.', () => {
    const page = shallow(<NewEvent {...testProps} />)
    const tree = shallowToJson(page)

    expect(tree).toMatchSnapshot()
  })
})
