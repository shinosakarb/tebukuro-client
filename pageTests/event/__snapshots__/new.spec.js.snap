// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewEvent renders the new event page. 1`] = `
<div>
  <h3>
    This is the event form page!
  </h3>
  <EventFormComponent
    onSubmit={[Function]}
  />
</div>
`;
